tiledb/sm/filter/webp_filter.cc

```diff
@@ -0,0 +1,233 @@
+ *
+ * This file implements class WebpFilter.
+ */
+
+#ifdef TILEDB_WEBP
+
+#include "tiledb/sm/filter/webp_filter.h"
+#include "tiledb/common/logger_public.h"
+#include "tiledb/common/status.h"
+#include "tiledb/sm/enums/filter_option.h"
+#include "tiledb/sm/filter/filter_buffer.h"
+#include "tiledb/sm/tile/tile.h"
+
+#include "webp/decode.h"
+#include "webp/encode.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+void WebpFilter::dump(FILE* out) const {
+  if (out == nullptr)
+    out = stdout;
+  fprintf(out, "WebpFilter");
+}
+
+Status WebpFilter::run_forward(
+    const Tile& tile,
+    Tile* const,
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  if (tile.type() != Datatype::UINT8) {
+    return LOG_STATUS(Status_FilterError("Unsupported input type"));
+  }
+  return run_forward(input_metadata, input, output_metadata, output);
+}
+
+Status WebpFilter::run_forward(
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  auto input_parts = input->buffers();
+  auto num_parts = static_cast<uint32_t>(input_parts.size());
+  uint32_t metadata_size = sizeof(uint32_t) + (num_parts * sizeof(uint32_t));
+  RETURN_NOT_OK(output_metadata->append_view(input_metadata));
+  RETURN_NOT_OK(output_metadata->prepend_buffer(metadata_size));
+  RETURN_NOT_OK(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  // TODO: Use X / Y dimension bounds for image dimensions
+  int width = 284, height = 178;
+  for (const auto& i : input_parts) {
+    auto data = static_cast<const uint8_t*>(i.data());
+    // Number of bytes encoded; Encoded result data buffer
+    size_t enc_size;
+    uint8_t* result;
+    switch (format_) {
+      case WebpInputFormat::WEBP_RGB:
+        enc_size =
+            WebPEncodeRGB(data, width, height, width * 3, quality_, &result);
+        break;
+      case WebpInputFormat::WEBP_RGBA:
+        enc_size =
+            WebPEncodeRGBA(data, width, height, width * 4, quality_, &result);
+        break;
+      case WebpInputFormat::WEBP_BGR:
+        enc_size =
+            WebPEncodeBGR(data, width, height, width * 3, quality_, &result);
+        break;
+      case WebpInputFormat::WEBP_BGRA:
+        enc_size =
+            WebPEncodeBGRA(data, width, height, width * 4, quality_, &result);
+        break;
+      case WebpInputFormat::WEBP_NONE:
+        return LOG_STATUS(Status_FilterError(
+            "Filter option TILEDB_FILTER_WEBP_FORMAT must be set"));
+    }
+
+    // Check if encoding failed
+    if (!enc_size) {
+      return LOG_STATUS(Status_FilterError("Error encoding image data"));
+    }
+
+    // Write encoded data to output buffer
+    RETURN_NOT_OK(output_metadata->write(&enc_size, sizeof(uint32_t)));
+    RETURN_NOT_OK(output->prepend_buffer(enc_size));
+    RETURN_NOT_OK(output->write(result, enc_size));
+    WebPFree(result);
```

Comment:
There will be a leak here if any of the three calls above return due to error. One option is to use `ScopedExecutor` to create an object which will execute `WebPFree` even if the function returns early.  eg

```
ScopedExecutor cleanup([](){
    if (enc_size > 0 && result != nullptr) { WebPFree(result); }
}
```